src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py

```diff
@@ -1,26 +1,574 @@
+                read_data_dict["Step Label"] = cls._get_step_label(line, read_mode)
+                continue
+
+            elif strp_line.startswith(WAVELENGTHS_KEY):
+                wavelengths = strp_line.split(":  ")
+                read_data_dict[WAVELENGTHS_KEY].extend(wavelengths[1].split(", "))
+                continue
+
+            line_data: list[str] = strp_line.split(",  ")
+            for read_datum in line_data:
+                splitted_datum = read_datum.split(": ")
+                if len(splitted_datum) != datum_len:
+                    continue
+                if splitted_datum[0] in list_keys:
+                    read_data_dict[splitted_datum[0]].append(splitted_datum[1])
+                else:
+                    read_data_dict[splitted_datum[0]] = splitted_datum[1]
+
+        return read_data_dict
+
+    @classmethod
+    def _get_step_label(cls, read_line: str, read_mode: str) -> Optional[str]:
+        read_data_len = 2
+        split_line = read_line.split("\t")
+        if len(split_line) != read_data_len:
+            msg = (
+                f"Expected the Read data line {split_line} to contain exactly 2 values."
+            )
             raise AllotropeConversionError(msg)
+        if split_line[1] != f"{read_mode.title()} Endpoint":
+            return split_line[1]
+
+        return None
+
+    @classmethod
+    def _get_filter_sets(
+        cls,
+        measurement_labels: list[str],
+        device_control_data: dict,
+        read_mode: ReadMode,
+    ) -> dict[str, FilterSet]:
+        filter_data: dict[str, FilterSet] = {}
+        if read_mode == ReadMode.ABSORBANCE:
+            return filter_data
+
+        emissions = device_control_data.get(EMISSION_KEY, [])
+        excitations = device_control_data.get(EXCITATION_KEY, [])
+        mirrors = device_control_data.get(MIRROR_KEY, [])
+        optics = device_control_data.get(OPTICS_KEY, [])
+        gains = device_control_data.get(GAIN_KEY, [])
+
+        for idx, label in enumerate(measurement_labels):
+            mirror = None
+            if mirrors and read_mode == ReadMode.FLUORESCENCE:
+                mirror = mirrors[idx]
+
+            filter_data[label] = FilterSet(
+                emission=emissions[idx],
+                gain=gains[idx],
+                excitation=excitations[idx] if excitations else None,
+                mirror=mirror,
+                optics=optics[idx] if optics else None,
+            )
+        return filter_data
+
+
+@dataclass(frozen=True)
+class LayoutData:
+    sample_identifiers: dict
+
+    @staticmethod
+    def create_default() -> LayoutData:
+        return LayoutData(sample_identifiers={})
+
+    @staticmethod
+    def create(layout_str: str) -> LayoutData:
+        identifiers = {}
+
+        layout_lines: list[str] = layout_str.splitlines()
+        # first line is "Layout", second line is column numbers
+        current_row = "A"
+        for i in range(2, len(layout_lines)):
```

Comment:
This data is in a tabular format, it may be better to read this in as a csv block using pandas?

Reply:
TBH, I inherited this from the old gen5 adapter and didn't think on changing it because it was working fine. Will look into it.

Reply:
Given the way the data is formatted (it can be multidimentional) and the fact that we would still have to iterate through all rows and columns to get the well positions. I think we wouldn't gain much by reading it as a csv with pandas.

Reply:
I think it would help with readability, as you should be able to get some column headers, but given that you inherited this code and this is already a pretty big PR, I'm good with leaving it 👍